src/utils/auth/auth.provider.tsx

```diff
@@ -0,0 +1,481 @@
+            );
+
+            return;
+          }
+
+          nextAuthRequests[completedAuthRequestIndex] = {
+            ...nextAuthRequests[completedAuthRequestIndex],
+            completedAt: Date.now(),
+            status: completedAuthRequestsStatus[i]
+          };
+
+          if (i === 0)
+            nextLastCompletedAuthRequest =
+              nextAuthRequests[completedAuthRequestIndex];
+        });
+
+        // Find the index of the next "pending" `AuthRequest`, or keep it unchanged if there are none left:
+        let nextCurrentAuthRequestIndex = currentAuthRequestIndex;
+
+        do {
+          nextCurrentAuthRequestIndex =
+            (nextCurrentAuthRequestIndex + 1) % nextAuthRequests.length;
+        } while (
+          nextCurrentAuthRequestIndex !== currentAuthRequestIndex &&
+          nextAuthRequests[nextCurrentAuthRequestIndex].status !== "pending"
+        );
+
+        if (nextCurrentAuthRequestIndex === currentAuthRequestIndex) {
+          nextCurrentAuthRequestIndex = nextAuthRequests.length;
+        }
+
+        return {
+          authRequests: nextAuthRequests,
+          currentAuthRequestIndex: nextCurrentAuthRequestIndex,
+          lastCompletedAuthRequest: nextLastCompletedAuthRequest
+        };
+      });
+    },
+    [authRequests]
+  );
+
+  useEffect(() => {
+    log(LOG_GROUP.AUTH, "Auth popup initialized. Waiting for AuthRequests");
+
+    isomorphicOnMessage("auth_request", (authRequestMessage) => {
+      log(LOG_GROUP.AUTH, "auth_request =", authRequestMessage);
+
+      // UnlockAuthRequests are not enqueued as those are simply used to open the popup to prompt users to enter their
+      // password and wait for the wallet to unlock:
+
+      if (!authRequestMessage?.data) {
+        console.warn("auth_request without data");
+        return;
+      }
+
+      const authRequest = authRequestMessage.data;
+
+      setAuthRequestContextState((prevAuthRequestContextState) => {
+        const {
+          authRequests,
+          currentAuthRequestIndex,
+          lastCompletedAuthRequest
+        } = prevAuthRequestContextState;
+
+        // TODO: Additional considerations when enqueueing new `AuthRequest`s:
+        //
+        // - `AuthRequest`s could be grouped by domain and/or tab. This means the auth popup could/should provide a
+        //   domain/tab selector and automatically select the active tab when the user switches it.
+        //
+        // - Separate `signDataItem`requests could/should automatically be combined into a single `batchSignDataItem`
+        //   (except maybe for the current `AuthRequest`, as otherwise the UI would constantly change as new requests
+        //   are added to the batch).
+
+        const nextAuthRequests = [
+          ...authRequests,
+          { ...authRequest, status: "pending" }
+        ] satisfies AuthRequest[];
+
+        // TODO: Add setting to decide whether we automatically jump to a new pending request when they arrive or stay
```

Comment:
👍 
IMO the default behavior should be to stay since dapps usually request Auth linearly as they need the items/txs signed

Reply:
True, but keep in mind the "current" one might have been selected manually by the user. Probably all we need is a visual indicator in the "selection bar" that a new one came in.